src/fabric.c

```diff
@@ -114,12 +119,34 @@ static int fi_register_provider(struct fi_provider *provider, void *dlhandle)
 		 * provider of the same name, so discard the
 		 * already-loaded one.
 		 */
-		FI_LOG(2, NULL, "an older %s provider was already loaded; keeping this one and ignoring the older one\n",
-			provider->name);
+		FI_LOG("fabric", FABRIC, LOG_INFO, 1,
+		       "an older %s provider was"
+		       "already loaded; keeping this one and ignoring the older"
+		       "one\n",
+		       provider->name);
 		cleanup_provider(prov->provider, prov->dlhandle);
 
 		prov->dlhandle = dlhandle;
 		prov->provider = provider;
+
+		/*
+		 * If context->internal[0] is not set then generate a position
+		 * to use. If this is the first one then use DYNAMIC_START, if
+		 * not then check the tail and add 1.
+		 */
+		if (!FI_LOG_CTX(provider)) {
+			FI_LOG_CTX(provider) = malloc(sizeof(int));
+			if (prov_tail) {
+				*(int *)FI_LOG_CTX(provider) =
+				    *(int *)FI_LOG_CTX(prov_tail->provider) + 1;
+			} else {
+				*(int *)FI_LOG_CTX(provider) = DYNAMIC_START;
+			}
```

Comment:
Yes, this is one of the portions I referred to when I said I had to do some cleanup. I'll do that soon.